solidity/contracts/middleware/FraudProofRouter.sol

```diff
@@ -0,0 +1,138 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
+pragma solidity >=0.6.11;
+
+/*@@@@@@@       @@@@@@@@@
+ @@@@@@@@@       @@@@@@@@@
+  @@@@@@@@@       @@@@@@@@@
+   @@@@@@@@@       @@@@@@@@@
+    @@@@@@@@@@@@@@@@@@@@@@@@@
+     @@@@@  HYPERLANE  @@@@@@@
+    @@@@@@@@@@@@@@@@@@@@@@@@@
+   @@@@@@@@@       @@@@@@@@@
+  @@@@@@@@@       @@@@@@@@@
+ @@@@@@@@@       @@@@@@@@@
+@@@@@@@@@       @@@@@@@@*/
+
+import {TypeCasts} from "../libs/TypeCasts.sol";
+import {FraudType, FraudMessage, Attribution} from "../libs/FraudMessage.sol";
+import {AttributeCheckpointFraud} from "../AttributeCheckpointFraud.sol";
+import {GasRouter} from "../client/GasRouter.sol";
+
+contract FraudProofRouter is GasRouter {
+    // ===================== State Variables =======================
+
+    AttributeCheckpointFraud public immutable attributeCheckpointFraud;
+
+    // store origin => signer => merkleTree => digest => {timestamp, fraudType}
+    mapping(uint32 origin => mapping(bytes32 signer => mapping(bytes32 merkleTree => mapping(bytes32 digest => Attribution))))
+        public fraudAttributions;
```

Comment:
this is pretty redundant with the internals of the `AttributeCheckpointFraud` contract

Reply:
the remote chain doesn't know the internals of AttributeCheckpointFraud and it becomes easier for the client sdk or contract.

Reply:
can we somehow merge these 2 contracts?

Reply:
it's possible